crypto-mcl.ts

```diff
@@ -0,0 +1,188 @@
+import {
+    waitReady as sodiumWaitReady,
+    crypto_box_keypair,
+    crypto_box_seal,
+    crypto_box_seal_open,
+    crypto_scalarmult_base,
+    crypto_secretbox_keygen,
+    from_string,
+    IKeyPair,
+    randombytes_buf,
+    to_base64,
+    to_string, crypto_secretbox_easy, crypto_secretbox_open_easy, secretbox_nonce
+} from "./sodium";
+import * as mcl from "./mcl";
+
+export async function waitReady(){
+    await mcl.waitReady();
+    await sodiumWaitReady();
+}
+
+/**
+ * What needs to be stored in the users' phone.
+ */
+interface IUserRecord {
+    R1: mcl.G2
+    Z1: mcl.GT
+    R2: mcl.G2
+    K: Uint8Array
+    c: Uint8Array
+    nonce: Uint8Array
+}
+
+/**
+ * This is the secret information of the location.
+ */
+interface IMasterTrace{
+    mskv: mcl.Fr;
+    info: string;
+    nonce: Uint8Array;
+    mskHAEnc: Uint8Array;
+}
+
+/**
+ * Things to print on QRcodes - the mtr part only needs to be used when an infection is signaled.
+ */
+interface ILocationData {
+    // entry information
+    ent: mcl.G2;
+    // proof for the entry information
+    pEnt: Uint8Array;
+    // master tracing information for the QRtrace)
+    mtr: IMasterTrace;
+}
+
+/**
+ * The request from the health authority to the location owner to create a proof to be sent to
+ * the visitors.
+ */
+interface IPreTrace{
+    Tprime: mcl.G1
+    info: string
+    nonce: Uint8Array
+    ctxt: Uint8Array
+}
+
+/**
+ * Section7 implements the cryptographic protocol in section 7 of the CrowdNotifier whitepaper.
+ */
+export class Section7{
+    /**
+     * Setup for the HealthAuthority - returns a keypair that can be used to encrypt the QRtrack codes.
+     */
+    static setupHA(): IKeyPair{
+        return crypto_box_keypair();
+    }
+
+    /**
+     * Creates the data necessary for a location owner.
+     * @param info is the public info of the location, e.g., name:location:room
+     * @param pkH the public key of the health authority
+     * @return the LocationData to be printed in 2 (3?) QRcodes
+     */
+    static genCode(info: string, pkH: Uint8Array): ILocationData{
+        const mskv = new mcl.Fr();
+        mskv.setByCSPRNG();
+        const mskHA = new mcl.Fr();
+        mskHA.setByCSPRNG();
+        const msk = mcl.add(mskv, mskHA);
```

Comment:
@wouterl I'd like to check that this is in the right Field (i.e., modulo the group order), and modulo operations are applied immediately rather than explicitly.

Reply:
@wouterl Fr is indeed the right thing to use, and no additional modulo operation needed